### PR TITLE
Moved tenant management code into a new package

### PR DESCRIPTION
**Pull request by hiranya911**
* Moved `TenantManager` and related classes into a new `multitenancy` package.
* Added `AuthHttpClient` and `FirebaseTenantClient` internal APIs for making REST calls.

### CODE REVIEW

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 1]
```diff
@@ -213,12 +214,7 @@ FirebaseTokenVerifier getSessionCookieVerifier(boolean checkRevoked) {
   }
 
   @Override
-  protected void doDestroy() {
-    // Only destroy the tenant manager if it has been created.
-    if (tenantManagerCreated.get()) {
-      getTenantManager().destroy();
-    }
-  }
+  protected void doDestroy() { }
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/FirebaseAuth.java:217]: Why don't we need to need to check if the app has been destroyed in the tenant manager?

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 2]
```diff
@@ -53,7 +54,7 @@
       @Override
       public TenantManager get() {
         tenantManagerCreated.set(true);
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/FirebaseAuth.java:56]: This variable is no longer needed now that doDestroy is a no-op.

File: src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      .put("CLAIMS_TOO_LARGE", "claims-too-large")
+      .put("CONFIGURATION_NOT_FOUND", "configuration-not-found")
+      .put("INSUFFICIENT_PERMISSION", "insufficient-permission")
+      .put("DUPLICATE_EMAIL", "email-already-exists")
+      .put("DUPLICATE_LOCAL_ID", "uid-already-exists")
+      .put("EMAIL_EXISTS", "email-already-exists")
+      .put("INVALID_CLAIMS", "invalid-claims")
+      .put("INVALID_EMAIL", "invalid-email")
+      .put("INVALID_PAGE_SELECTION", "invalid-page-token")
+      .put("INVALID_PHONE_NUMBER", "invalid-phone-number")
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java:65]: Did you purposefully omit some error codes from this map? e.g. TENANT_ID_MISMATCH

File: src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      response = request.execute();
+      return response.parseAs(clazz);
+    } catch (HttpResponseException e) {
+      // Server responded with an HTTP error
+      handleHttpError(e);
+      return null;
+    } catch (IOException e) {
+      // All other IO errors (Connection refused, reset, parse error etc.)
+      throw new FirebaseAuthException(
+          INTERNAL_ERROR, "Error while calling user management backend service", e);
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java:131]: Should we use a broader term than "user management backend service" if this is being used for tenant/provider-related requests too?

File: src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+      }
+    }
+  }
+
+  private void handleHttpError(HttpResponseException e) throws FirebaseAuthException {
+    try {
+      HttpErrorResponse response = jsonFactory.fromString(e.getContent(), HttpErrorResponse.class);
+      String code = ERROR_CODES.get(response.getErrorCode());
+      if (code != null) {
+        throw new FirebaseAuthException(code, "User management service responded with an error", e);
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java:148]: Should we use a broader term than "User management service" if this is being used for tenant/provider-related requests too?

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java [hunk 6]
```diff
@@ -71,37 +61,10 @@
  */
 class FirebaseUserManager {
 
-  static final String CONFIGURATION_NOT_FOUND_ERROR = "configuration-not-found";
-  static final String TENANT_ID_MISMATCH_ERROR = "tenant-id-mismatch";
-  static final String TENANT_NOT_FOUND_ERROR = "tenant-not-found";
   static final String USER_NOT_FOUND_ERROR = "user-not-found";
   static final String INTERNAL_ERROR = "internal-error";
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/FirebaseUserManager.java:65]: Wouldn't it be nice to keep all of these constants in one place? AuthHttpClient is a good spot to keep them. And it appears INTERNAL_ERROR is currently being duplicated both here and AuthHttpClient.

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java [hunk 7]
```diff
@@ -449,27 +367,12 @@ ListSamlProviderConfigsResponse listSamlProviderConfigs(int maxResults, String p
...
-        }
-      } else {
-        maskBuilder.add(entry.getKey());
-      }
-    }
-    return maskBuilder.build();
+    httpClient.sendRequest("DELETE", url, null, GenericJson.class);
   }
 
   private static String getTenantUrlSuffix(String tenantId) {
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/FirebaseUserManager.java:378]: Can `getTenantUrlSuffix` be removed now that the tenant-related methods have been removed?

File: src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import com.google.firebase.internal.ApiClientUtils;
+import java.util.Map;
+
+final class FirebaseTenantClient {
+
+  static final int MAX_LIST_TENANTS_RESULTS = 100;
+
+  static final String TENANT_NOT_FOUND_ERROR = "tenant-not-found";
+
+  static final String INTERNAL_ERROR = "internal-error";
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java:44]: Consolidate these errors in AuthHttpClient? INTERNAL_ERROR is currently being duplicated across both.

File: src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    httpClient.sendRequest("DELETE", url, null, GenericJson.class);
+  }
+
+  ListTenantsResponse listTenants(int maxResults, String pageToken)
+      throws FirebaseAuthException {
+    ImmutableMap.Builder<String, Object> builder =
+        ImmutableMap.<String, Object>builder().put("pageSize", maxResults);
+    if (pageToken != null) {
+      checkArgument(!pageToken.equals(
+          ListTenantsPage.END_OF_LIST), "Invalid end of list page token");
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java:97]: Add period after "token".

File: src/main/java/com/google/firebase/auth/multitenancy/ListTenantsPage.java [hunk 10]
```diff
@@ -198,9 +197,9 @@ ListTenantsResponse fetch(int maxResults, String pageToken)
 
   static class DefaultTenantSource implements TenantSource {
 
-    private final FirebaseUserManager userManager;
+    private final FirebaseTenantClient userManager;
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/multitenancy/ListTenantsPage.java:200]: Rename this variable to `tenantClient`. It's not a FirebaseUserManager anywmore!

File: src/main/java/com/google/firebase/auth/multitenancy/TenantManager.java [hunk 11]
```diff
@@ -14,45 +14,54 @@
...
-  private final Object lock = new Object();
-  private final AtomicBoolean destroyed = new AtomicBoolean(false);
-
   private final FirebaseApp firebaseApp;
-  private final FirebaseUserManager userManager;
+  private final FirebaseTenantClient tenantClient;
   private final Map<String, TenantAwareFirebaseAuth> tenantAwareAuths;
 
-  TenantManager(FirebaseApp firebaseApp, FirebaseUserManager userManager) {
+  /**
```
Comment (micahstairs) [inline src/main/java/com/google/firebase/auth/multitenancy/TenantManager.java:50]: This should be `{@link TenantManager}`.

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java [hunk 12]
```diff
@@ -78,11 +82,13 @@
...
           .setAndroidMinimumVersion("6")
           .build();
+
   private static final Map<String, Object> ACTION_CODE_SETTINGS_MAP =
           ACTION_CODE_SETTINGS.getProperties();
 
   private static final String PROJECT_BASE_URL =
       "https://identitytoolkit.googleapis.com/v2/projects/test-project-id";
+
   private static final String TENANTS_BASE_URL = PROJECT_BASE_URL + "/tenants";
```
Comment (micahstairs) [inline src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java:92]: Is `TENANTS_BASE_URL` still needed?

File: src/test/java/com/google/firebase/auth/ProviderConfigTestUtils.java [hunk 13]
```diff
@@ -26,68 +26,72 @@
 import java.util.concurrent.ExecutionException;
 import org.junit.rules.ExternalResource;
 
-class ProviderConfigTestUtils {
+public class ProviderConfigTestUtils {
 
-  static void assertOidcProviderConfigDoesNotExist(
+  static final String CONFIGURATION_NOT_FOUND_ERROR = "configuration-not-found";
```
Comment (micahstairs) [inline src/test/java/com/google/firebase/auth/ProviderConfigTestUtils.java:31]: Consolidate in AuthHttpClient?

File: src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java [hunk 14]
```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright 2020 Google Inc.
```
Comment (micahstairs) [inline src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java:2]: LLC

File: src/test/java/com/google/firebase/auth/multitenancy/TenantAwareFirebaseAuthIT.java [hunk 15]
```diff
@@ -247,7 +258,7 @@ public void testVerifyTokenWithWrongTenantAwareClient() throws Exception {
       fail("No error thrown for verifying a token with the wrong tenant-aware client");
     } catch (ExecutionException e) {
       assertTrue(e.getCause() instanceof FirebaseAuthException);
-      assertEquals(FirebaseUserManager.TENANT_ID_MISMATCH_ERROR,
+      Assert.assertEquals("tenant-id-mismatch",
```
Comment (micahstairs) [inline src/test/java/com/google/firebase/auth/multitenancy/TenantAwareFirebaseAuthIT.java:261]: Use a static import for `Assert.assertEquals`?

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java (hunk 1 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/FirebaseAuth.java:217]: I think this is something we ought to get rid of from the entire codebase. `FirebaseAuth` is the only class that does this today. I think it's fair to not do it in any new classes.

File: src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java (hunk 3 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java:65]: Yes. That's not a real error code sent by the Auth backend.

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java (hunk 6 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/FirebaseUserManager.java:65]: Moved the remaining 2 error codes to `AuthHttpClient`

File: src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java (hunk 12 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/FirebaseUserManagerTest.java:92]: There are tests in this file that makes tenant-aware user management calls.

File: src/test/java/com/google/firebase/auth/multitenancy/TenantAwareFirebaseAuthIT.java (hunk 15 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/multitenancy/TenantAwareFirebaseAuthIT.java]: Done

File: src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java (hunk 14 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/multitenancy/FirebaseTenantClientTest.java]: Done

File: src/test/java/com/google/firebase/auth/ProviderConfigTestUtils.java (hunk 13 above)
Comment (hiranya911) [reply src/test/java/com/google/firebase/auth/ProviderConfigTestUtils.java]: Done

File: src/main/java/com/google/firebase/auth/multitenancy/TenantManager.java (hunk 11 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/multitenancy/TenantManager.java:50]: Done

File: src/main/java/com/google/firebase/auth/multitenancy/ListTenantsPage.java (hunk 10 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/multitenancy/ListTenantsPage.java]: Done

File: src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java (hunk 9 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java]: Done

File: src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java (hunk 8 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/multitenancy/FirebaseTenantClient.java]: Done

File: src/main/java/com/google/firebase/auth/FirebaseUserManager.java (hunk 7 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/FirebaseUserManager.java]: Done

File: src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java (hunk 5 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java]: Done

File: src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java (hunk 4 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java]: Good point. Done.

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java (hunk 2 above)
Comment (hiranya911) [reply src/main/java/com/google/firebase/auth/FirebaseAuth.java]: Done

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java (hunk 1 above)
Comment (micahstairs) [reply src/main/java/com/google/firebase/auth/FirebaseAuth.java:214]: Ack.

File: src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java (hunk 3 above)
Comment (micahstairs) [reply src/main/java/com/google/firebase/auth/internal/AuthHttpClient.java:68]: Ack.

**Pull request merged**